File: web/js/components/notes.js
async function getNotes(folder) {
  let notes = await eel.get_notes(folder)();
  let formAction = `sendTemplate({}, 'create', 'notes', 'main'); getNoteSide('loadFolder()');`;
  let html = await loadTemplate({ 'formAction': formAction, 'title': 'notes', 'icon': 'bi bi-journal-plus'}, 'sectionHeader', 'folders');
  html += await eel.render_template_list(notes, 'notes', 'items')();
  document.getElementById('notes').innerHTML = html;
}

async function createNote() {
  const title = document.getElementById('noteName').value;
  const content = document.getElementById('noteContent').value;
  const folder = getCookie('folder');
  id = await eel.create_note(title, content, folder)();
  await getNote(id);
  setMainItem('note', id);
}

async function getNoteSide(callback, id) {
  let notes = await eel.get_notes(getCookie('folder'))();
  if (id) {
    for (let i = 0; i < notes.length; i++) {
      notes[i].id = notes[i].id + 'i'
      if (notes[i].id == id + 'i') {
        notes.splice(i, 1);
      }
    }
  }
  console.log(notes)
  let html = `
    <div class="row">
      <h3 class="align row center">other notes</h3>
    </div>
`;
  html += await eel.render_template_list(notes, 'notes', 'items')();
  document.getElementById('notesSide').innerHTML = html;
}

async function getNote(id) {
  if (typeof id == 'string') {
    if(id.includes('i')) {
      var strippedString = id.slice(0, -1);
      id = parseInt(strippedString, 10);
    }
    id = parseInt(id, 10);
  }
  let note = await eel.get_note(id)();
  setMainItem('note', id);
  noteContent = await replaceMarkdownLinks(note.content);
  await sendTemplate({ 'id': id, 'name': note.name, 'content': marked.parse(noteContent), 'time': note.time }, 'page', 'notes', 'main');
  await processTasks('main');
  await getNoteSide('getNote(' + id + ')', id);
}

async function editNote(id) {
  note = await eel.get_note(id)();
  await sendTemplate({ 'id': id, 'name': note.name, 'content': note.content}, 'edit', 'notes', 'main');
  await getNoteSide('getNote(' + id + ')', id);
  setMainItem('note', id);
}

async function saveNote(id) {
  const title = document.getElementById('noteName').value;
  const content = document.getElementById('noteContent').value;
  const folder = getCookie('folder');
  await eel.update_note(id, title, content, folder)();
  await getNote(id);
}

async function deleteNote(id) {
  await eel.delete_note(id)();
  await loadFolder();
}

File: api/notes.py
import sys
sys.path.append("..")
import eel
from models import Note
from api.attachments import delete_all_attachments, delete_attachment

# notes api


@eel.expose
def get_notes(folder_id):
    notes = Note.select().where(Note.folder_id == folder_id)
    return [{'id': note.id, 'name': note.name, 'time': note.time} for note in notes]


@eel.expose
def get_note(id):
    note = Note.get(Note.id == id)
    return {'name': note.name, 'content': note.content, 'time': note.time}


@eel.expose
def delete_note(id):
    note = Note.get(Note.id == id)
    delete_all_attachments(note.id, 'note')
    return note.delete_instance()


@eel.expose
def create_note(name, content, folder_id):
    note = Note.create(name=name, content=content, folder=folder_id)
    return note.id


@eel.expose
def update_note(note_id, name, content, folder_id):
    note = Note.get(Note.id == note_id)
    note.name = name
    note.content = content
    if folder_id != '':
        note.folder = folder_id
    note.save()
